### PR TITLE
API/OP error message enhancement

### PR DESCRIPTION
**Pull request by 0YuanZhang0**
1. gaussian_random：python && c++ 
2. precision_recall：c++ 
3. sequence_concat：python && c++ 
4. sequence_pad：python && c++ 
5. sequence_reshape：python && c++ 
6. sequence_reverse：python && c++ 
7. sequence_unpad：python && c++ 
8. truncated_gaussian_random：c++ 
9. gaussian_random_batch_size_like：python

### CODE REVIEW

File: paddle/fluid/operators/gaussian_random_op.cc [hunk 1]
```diff
@@ -50,16 +50,21 @@ class GaussianRandomOp : public framework::OperatorWithKernel {
   using framework::OperatorWithKernel::OperatorWithKernel;
 
   void InferShape(framework::InferShapeContext* ctx) const override {
-    PADDLE_ENFORCE(ctx->HasOutput("Out"),
-                   "Output(Out) of GaussianRandomOp should not be null.");
+    PADDLE_ENFORCE_EQ(
+        ctx->HasOutput("Out"), true,
+        platform::errors::InvalidArgument(
```
Comment (chenwhql) [inline paddle/fluid/operators/gaussian_random_op.cc]: `InvalidArgument` -> `NotFound`, or use MACRO `OP_INOUT_CHECK`

File: paddle/fluid/operators/gaussian_random_op.cc [hunk 2]
```diff
@@ -50,16 +50,21 @@ class GaussianRandomOp : public framework::OperatorWithKernel {
...
     std::vector<int64_t> temp;
     temp.reserve(shape.size());
     for (auto dim : shape) {
       temp.push_back(static_cast<int64_t>(dim));
     }
-    PADDLE_ENFORCE(shape.size() > 0UL,
-                   "shape can be one int or array. shape must be set.");
+    PADDLE_ENFORCE_GT(shape.size(), 0UL,
+                      platform::errors::InvalidArgument(
+                          "shape size should be int or array, and larger than "
```
Comment (chenwhql) [inline paddle/fluid/operators/gaussian_random_op.cc]: shape size can be array?

File: paddle/fluid/operators/metrics/precision_recall_op.cc [hunk 3]
```diff
@@ -22,57 +22,73 @@ class PrecisionRecallOp : public framework::OperatorWithKernel {
...
-    PADDLE_ENFORCE(ctx->HasInput("Labels"),
-                   "Input(Labels) should not be null.");
-    PADDLE_ENFORCE(ctx->HasOutput("BatchMetrics"),
-                   "Output(BatchMetrics) should not be null.");
-    PADDLE_ENFORCE(ctx->HasOutput("AccumMetrics"),
-                   "Output(AccumMetrics) should not be null.");
-    PADDLE_ENFORCE(ctx->HasOutput("AccumStatesInfo"),
-                   "Output(AccumStatesInfo) should not be null.");
+    PADDLE_ENFORCE_EQ(ctx->HasInput("MaxProbs"), true,
+                      platform::errors::InvalidArgument(
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/precision_recall_op.cc]: `InvalidArgument` -> `NotFound`, or use MACRO `OP_INOUT_CHECK` , same for others

File: paddle/fluid/operators/metrics/precision_recall_op.cc [hunk 4]
```diff
@@ -22,57 +22,73 @@ class PrecisionRecallOp : public framework::OperatorWithKernel {
...
+                          "Input(MaxProbs) should not be null."));
+    PADDLE_ENFORCE_EQ(ctx->HasInput("Indices"), true,
+                      platform::errors::InvalidArgument(
+                          "Input(Indices) should not be null."));
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/precision_recall_op.cc]: who's Input? same for others

File: paddle/fluid/operators/metrics/precision_recall_op.cc [hunk 5]
```diff
@@ -22,57 +22,73 @@ class PrecisionRecallOp : public framework::OperatorWithKernel {
...
+              max_probs_dims[1]));
+      PADDLE_ENFORCE_EQ(ctx->GetInputDim("Indices"), max_probs_dims,
+                        platform::errors::InvalidArgument(
+                            "The shape of Input(Indices) should bes same with "
+                            "max_probs_dims, But received (%d) != (%d)",
+                            ctx->GetInputDim("Indices"), max_probs_dims));
+      PADDLE_ENFORCE_EQ(max_probs_dims[0], labels_dims[0],
+                        platform::errors::InvalidArgument(
+                            "The 1st dimension of Input(MaxProbs) and "
+                            "Input(Labels) both are batch_size and the shape "
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/precision_recall_op.cc]: both should be

File: paddle/fluid/operators/metrics/precision_recall_op.cc [hunk 6]
```diff
@@ -22,57 +22,73 @@ class PrecisionRecallOp : public framework::OperatorWithKernel {
...
 
       if (ctx->IsRuntime()) {
-        PADDLE_ENFORCE_EQ(weights_dims,
-                          framework::make_ddim({max_probs_dims[0], 1}),
-                          "The shape of Input(Weights) should be "
-                          "[batch_size, 1].");
+        PADDLE_ENFORCE_EQ(
+            weights_dims, framework::make_ddim({max_probs_dims[0], 1}),
+            platform::errors::InvalidArgument(
+                "The shape of Input(Weights) should be [batch_size, 1]."));
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/precision_recall_op.cc]: but received what?

File: paddle/fluid/operators/metrics/precision_recall_op.cc [hunk 7]
```diff
@@ -22,57 +22,73 @@ class PrecisionRecallOp : public framework::OperatorWithKernel {
...
       auto states_dims = ctx->GetInputDim("StatesInfo");
 
       if (ctx->IsRuntime()) {
-        PADDLE_ENFORCE_EQ(states_dims, framework::make_ddim({cls_num, 4}),
-                          "The shape of Input(StatesInfo) should be "
-                          "[class_number, 4].");
+        PADDLE_ENFORCE_EQ(
+            states_dims, framework::make_ddim({cls_num, 4}),
+            platform::errors::InvalidArgument(
+                "The shape of Input(StatesInfo) should be [class_number, 4]."));
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/precision_recall_op.cc]: same above

File: paddle/fluid/operators/metrics/precision_recall_op.cc [hunk 8]
```diff
@@ -22,57 +22,73 @@ class PrecisionRecallOp : public framework::OperatorWithKernel {
...
+                        platform::errors::InvalidArgument(
+                            "The shape of Input(Indices) should bes same with "
+                            "max_probs_dims, But received (%d) != (%d)",
+                            ctx->GetInputDim("Indices"), max_probs_dims));
+      PADDLE_ENFORCE_EQ(max_probs_dims[0], labels_dims[0],
+                        platform::errors::InvalidArgument(
+                            "The 1st dimension of Input(MaxProbs) and "
+                            "Input(Labels) both are batch_size and the shape "
+                            "should be the same. But received (%d) != (%d)",
+                            max_probs_dims[0], labels_dims[0]));
```
Comment (xyzhou-puck) [inline paddle/fluid/operators/metrics/precision_recall_op.cc]: please make the error information more clear, 4 != 5 is not accurate enough for me.
you could use "but received max_probs_dims = %d, labels_dims = %d".

File: paddle/fluid/operators/sequence_ops/sequence_concat_op.h [hunk 9]
```diff
@@ -73,16 +73,23 @@ class SeqConcatKernel : public framework::OpKernel<T> {
...
+                              "The number of sequence must be same between "
+                              "each input. But received (%d) != (%d)",
+                              lod_size, x.get().lod()[0].size()));
       }
     }
-    PADDLE_ENFORCE_NE(lod_size, 0, "Each input must have sequence information");
+    PADDLE_ENFORCE_NE(lod_size, 0,
+                      platform::errors::InvalidArgument(
+                          "Each input must have sequence information. But "
+                          "received input lod size is(%d)",
```
Comment (chenwhql) [inline paddle/fluid/operators/sequence_ops/sequence_concat_op.h]: user may be confused about the relation between `sequence info` and `lod size`, we can just tell users no sequence information here

File: paddle/fluid/operators/sequence_ops/sequence_pad_op.cc [hunk 10]
```diff
@@ -54,31 +64,43 @@ class SequencePadOp : public framework::OperatorWithKernel {
...
       if (padded_length == -1) {
         padded_length = max_seq_len;
       }
-      PADDLE_ENFORCE_GE(padded_length, max_seq_len,
-                        "The Attr(padded_length) must be -1 or an int greater "
-                        "than the length of the longest original sequence.");
+      PADDLE_ENFORCE_GE(
+          padded_length, max_seq_len,
+          platform::errors::InvalidArgument(
+              "The Attr(padded_length) must be -1 or an int greater than the "
```
Comment (chenwhql) [inline paddle/fluid/operators/sequence_ops/sequence_pad_op.cc]: an int -> an int value?

File: paddle/fluid/operators/sequence_ops/sequence_reshape_op.h [hunk 11]
```diff
@@ -56,10 +62,11 @@ class SequenceReshapeKernel : public framework::OpKernel<T> {
         size_t offset = 0;
         offset = (seq_len * in_width) / out_width;
         PADDLE_ENFORCE_EQ(offset * out_width, seq_len * in_width,
-                          "Please make sure (sequence_length * dimension) can "
-                          "be divided by new_dim with no remainder for each "
-                          "sequence. The %dth sequence is invalid.",
-                          i + 1);
+                          platform::errors::InvalidArgument(
+                              "Please make sure (sequence_length * dimension) "
+                              "can be divided by new_dim with no remainder for "
```
Comment (chenwhql) [inline paddle/fluid/operators/sequence_ops/sequence_reshape_op.h]: `new_dim` means what here?

File: paddle/fluid/operators/metrics/precision_recall_op.h [hunk 12]
```diff
@@ -58,11 +58,25 @@ class PrecisionRecallKernel : public framework::OpKernel<T> {
...
 
-      PADDLE_ENFORCE(idx >= 0 && idx < cls_num,
-                     "Class index of each instance should be in "
-                     "[0, class_number).");
-      PADDLE_ENFORCE(label >= 0 && label < cls_num,
-                     "Label of each instance should be in [0, class_number).");
+      PADDLE_ENFORCE_GE(idx, 0, platform::errors::InvalidArgument(
+                                    "Class index of each instance should be "
+                                    "larger than 0, But received (%d)",
+                                    idx));
```
Comment (xyzhou-puck) [inline paddle/fluid/operators/metrics/precision_recall_op.h]: the class index of each instance should be not less than 0

Comment (xyzhou-puck) [reply paddle/fluid/operators/metrics/precision_recall_op.h]: please double check GE and other ones.

File: paddle/fluid/operators/truncated_gaussian_random_op.cc [hunk 13]
```diff
@@ -182,17 +182,21 @@ class TruncatedGaussianRandomOp : public framework::OperatorWithKernel {
...
     std::vector<int64_t> out_dim;
     out_dim.reserve(shape.size());
     for (auto dim : shape) {
       out_dim.push_back(static_cast<int64_t>(dim));
     }
-    PADDLE_ENFORCE(shape.size() > 0UL,
-                   "shape can be one int or array. shape must be set.");
+    PADDLE_ENFORCE_GT(shape.size(), 0UL,
+                      platform::errors::InvalidArgument(
+                          "shape can be one int or array. shape must be set, "
```
Comment (chenwhql) [inline paddle/fluid/operators/truncated_gaussian_random_op.cc]: can we directly tell use `shape is not set` or `shape is error` firstly?

**Review by lanxianghit — APPROVED** [review]

File: paddle/fluid/operators/metrics/precision_recall_op.cc (hunk 8 above)
Comment (0YuanZhang0) [reply paddle/fluid/operators/metrics/precision_recall_op.cc]: 已修改

File: paddle/fluid/operators/metrics/precision_recall_op.h (hunk 12 above)
Comment (0YuanZhang0) [reply paddle/fluid/operators/metrics/precision_recall_op.h]: 已修改

File: python/paddle/fluid/layers/sequence_lod.py [hunk 14]
```diff
@@ -1001,6 +1014,10 @@ def sequence_unpad(x, length, name=None):
     assert not in_dygraph_mode(), (
         "sequence layer is not supported in dygraph mode yet.")
     helper = LayerHelper('sequence_unpad', input=x, **locals())
+    check_variable_and_dtype(x, 'x', ['float32', 'float64', 'int32', 'int64'],
+                             'fluid.layers.sequence_unpad')
+    check_variable_and_dtype(length, 'length', ['int64'],
```
Comment (zhangting2020) [inline python/paddle/fluid/layers/sequence_lod.py:1019]: sequence_unpad的`length`添加类型检查，限制为int64的类型，导致content_dnn模型挂了。

```
  File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/layers/sequence_lod.py", line 1020, in sequence_unpad
    'fluid.layers.sequence_unpad')
  File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/data_feeder.py", line 80, in check_variable_and_dtype
    check_dtype(input.dtype, input_name, expected_dtype, op_name, extra_message)
  File "/usr/local/lib/python2.7/dist-packages/paddle/fluid/data_feeder.py", line 115, in check_dtype
    extra_message))
TypeError: The data type of 'length' in fluid.layers.sequence_unpad must be ['int64'], but received float32. 
```
这个参数是个var，在模型中是sequence_pad输出的。经排查，发现是sequence_pad的python API中将输出`Length`的类型设置为输入的类型。输入是float32，就会将length设置为float32类型（尽管OP的c++实现中是int64）。由于输入类型检查是compile time进行的，导致sequence_unpad报错。
https://github.com/PaddlePaddle/Paddle/blob/7dac3226ec0d882883a7661e07d5ccb4c6521460/python/paddle/fluid/layers/sequence_lod.py#L942-L944

sequence_pad Python端应该将输出length设置为int64，与c++端一致。

**Pull request merged**